Unsuported "priority" option for vim v8.0 | vim.error: Vim(sign):E474: Invalid argument

After some investigation and troubleshooting, I can see that the `priority` option is being used along with the `sign` command in `vim`, which introduces an error in versions where that option was not added yet (e.g. v.8.0)

If you try the following command[a] in `vim` 8.0 you'll see that it fails with the error: **E474: Invalid argument**:

```
:sign place 6145 name=current priority=99 line=4 file=/tmp/xdebugtest.php
```

Nonetheless, the same thing works just fine in `neovim` or newer versions of `vim` (e.g. 8.2)

This parameter was introduced in this commit: [`348bfee`](https://github.com/vim-vdebug/vdebug/commit/348bfeefb08cd8f081d3cd0d2838d6bb8e0d4b11#diff-2cf591b434babfdc22d874c37a8e1b48), but at least one more also make use of it: [`97e637c`](https://github.com/vim-vdebug/vdebug/commit/97e637cacd166839c975ba5a9485c549cc180540)

After updating I started receiving the following error (extracted from `vdebug.log`) just after I start a debugging session and a breakpoint is reached[1]. I have included some additional information just in case it's relevant[2].

Note: as my OS version is still in LTS, many people might face this issue.

Regards,

[1]
```
...
- [Debug] {vie 10 2020 13:20:36} keymapper: reload_keys
- [Info] {vie 10 2020 13:20:36} Found connection from ('192.168.56.1', 53606)
- [Debug] {vie 10 2020 13:20:36} Command: context_names -i 1
- [Debug] {vie 10 2020 13:20:36} Response: <?xml version="1.0" encoding="iso-8859-1"?>
<response xmlns="urn:debugger_protocol_v1" xmlns:xdebug="https://xdebug.org/dbgp/xdebug" command="context_names" transaction_id="1"><context name="Locals" id="0"></context><context name="Superglobals" id="1"></context><context name="User defined constants" id="2"></context></response>
- [Debug] {vie 10 2020 13:20:36} Available context names: {0: 'Locals', 1: 'Superglobals', 2: 'User defined constants'}
- [Info] {vie 10 2020 13:20:36} Updated line number of breakpoint 11000 to 4
- [Info] {vie 10 2020 13:20:36} Registering 1 breakpoints with the debugger
- [Debug] {vie 10 2020 13:20:36} Command: breakpoint_set -i 2 -t line -f "file:///tmp/xdebugtest.php" -n 4 -s enabled
- [Debug] {vie 10 2020 13:20:36} Response: <?xml version="1.0" encoding="iso-8859-1"?>
<response xmlns="urn:debugger_protocol_v1" xmlns:xdebug="https://xdebug.org/dbgp/xdebug" command="breakpoint_set" transaction_id="2" state="enabled" id="250870001"></response>
- [Debug] {vie 10 2020 13:20:36} starting with step_into (break_on_open = 1)
- [Debug] {vie 10 2020 13:20:36} Command: step_into -i 3
- [Debug] {vie 10 2020 13:20:36} Response: <?xml version="1.0" encoding="iso-8859-1"?>
<response xmlns="urn:debugger_protocol_v1" xmlns:xdebug="https://xdebug.org/dbgp/xdebug" command="step_into" transaction_id="3" status="break" reason="ok"><xdebug:message filename="file:///tmp/xdebugtest.php" lineno="4"></xdebug:message></response>
- [Debug] {vie 10 2020 13:20:36} refresh event
- [Info] {vie 10 2020 13:20:36} Dispatching refresh event
- [Debug] {vie 10 2020 13:20:36} ** RefreshEvent
- [Info] {vie 10 2020 13:20:36} Getting stack information
- [Debug] {vie 10 2020 13:20:36} keymapper: reload_keys
- [Debug] {vie 10 2020 13:20:36} Command: stack_get -i 4
- [Debug] {vie 10 2020 13:20:36} Response: <?xml version="1.0" encoding="iso-8859-1"?>
<response xmlns="urn:debugger_protocol_v1" xmlns:xdebug="https://xdebug.org/dbgp/xdebug" command="stack_get" transaction_id="4"><stack where="{main}" level="0" type="file" filename="file:///tmp/xdebugtest.php" lineno="4"></stack></response>
- [Info] {vie 10 2020 13:20:36} Moving to current position in source window
- [Info] {vie 10 2020 13:20:36} Setting source file: /tmp/xdebugtest.php
- [Debug] {vie 10 2020 13:20:36} Setting source line number: 4
- [Info] {vie 10 2020 13:20:36} Placing pointer sign on line 4
- [ERROR] {vie 10 2020 13:20:36} An error occured: <class 'vim.error'>
Traceback (most recent call last):
  File "/home/jespinal/.vim/pluggin/vdebug/python3/vdebug/event.py", line 781, in dispatch_event
    Dispatcher.events[name](self.__session_handler).run(*args)
  File "/home/jespinal/.vim/pluggin/vdebug/python3/vdebug/event.py", line 306, in run
    self.session.cur_lineno)
  File "/home/jespinal/.vim/pluggin/vdebug/python3/vdebug/ui/vimui.py", line 236, in set_source_position
    self.sourcewin.place_pointer(lineno)
  File "/home/jespinal/.vim/pluggin/vdebug/python3/vdebug/ui/vimui.py", line 392, in place_pointer
    % (self.pointer_sign_id, line, self.file))
vim.error: Vim(sign):E474: Invalid argument```
I got sure to remove all other plugins in order to discard any conflicts, but the same behavior is observed.
```

[2] Additional info

```
$ cat /etc/issue
Ubuntu 18.04.4 LTS \n \l
```

```
$ which python3
/usr/bin/python3
$ python3 --version
Python 3.6.9
```

```
$ vim --version
VIM - Vi IMproved 8.0 (2016 Sep 12, compiled Mar 18 2020 18:29:15)
Included patches: 1-1453
Modified by pkg-vim-maintainers@lists.alioth.debian.org
Compiled by pkg-vim-maintainers@lists.alioth.debian.org
Huge version with GTK2 GUI.  Features included (+) or not (-):
+acl               +farsi             +mouse_sgr         -tag_any_white
+arabic            +file_in_path      -mouse_sysmouse    +tcl
+autocmd           +find_in_path      +mouse_urxvt       +termguicolors
-autoservername    +float             +mouse_xterm       +terminal
+balloon_eval      +folding           +multi_byte        +terminfo
+balloon_eval_term -footer            +multi_lang        +termresponse
+browse            +fork()            -mzscheme          +textobjects
++builtin_terms    +gettext           +netbeans_intg     +timers
+byte_offset       -hangul_input      +num64             +title
+channel           +iconv             +packages          +toolbar
+cindent           +insert_expand     +path_extra        +user_commands
+clientserver      +job               +perl              +vertsplit
+clipboard         +jumplist          +persistent_undo   +virtualedit
+cmdline_compl     +keymap            +postscript        +visual
+cmdline_hist      +lambda            +printer           +visualextra
+cmdline_info      +langmap           +profile           +viminfo
+comments          +libcall           -python            +vreplace
+conceal           +linebreak         +python3           +wildignore
+cryptv            +lispindent        +quickfix          +wildmenu
+cscope            +listcmds          +reltime           +windows
+cursorbind        +localmap          +rightleft         +writebackup
+cursorshape       +lua               +ruby              +X11
+dialog_con_gui    +menu              +scrollbind        -xfontset
+diff              +mksession         +signs             +xim
+digraphs          +modify_fname      +smartindent       +xpm
+dnd               +mouse             +startuptime       +xsmp_interact
-ebcdic            +mouseshape        +statusline        +xterm_clipboard
+emacs_tags        +mouse_dec         -sun_workshop      -xterm_save
+eval              +mouse_gpm         +syntax            
+ex_extra          -mouse_jsbterm     +tag_binary        
+extra_search      +mouse_netterm     +tag_old_static    
   system vimrc file: "$VIM/vimrc"
     user vimrc file: "$HOME/.vimrc"
 2nd user vimrc file: "~/.vim/vimrc"
      user exrc file: "$HOME/.exrc"
  system gvimrc file: "$VIM/gvimrc"
    user gvimrc file: "$HOME/.gvimrc"
2nd user gvimrc file: "~/.vim/gvimrc"
       defaults file: "$VIMRUNTIME/defaults.vim"
    system menu file: "$VIMRUNTIME/menu.vim"
  fall-back for $VIM: "/usr/share/vim"
Compilation: gcc -c -I. -Iproto -DHAVE_CONFIG_H -DFEAT_GUI_GTK  -pthread -I/usr/include/gtk-2.0 -I/usr/lib/x86_64-linux-gnu/gtk-2.0/include -I/usr/include/gio-unix-2.0/ -I/usr/include/cairo -I/usr/include/pango-1.0 -I/usr/include/atk-1.0 -I/usr/include/cairo -I/usr/include/pixman-1 -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/libpng16 -I/usr/include/pango-1.0 -I/usr/include/harfbuzz -I/usr/include/pango-1.0 -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/freetype2 -I/usr/include/libpng16 -Wdate-time  -g -O2 -fdebug-prefix-map=/build/vim-dfCUyu/vim-8.0.1453=. -fstack-protector-strong -Wformat -Werror=format-security -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1       
Linking: gcc   -L. -Wl,-Bsymbolic-functions -Wl,-z,relro -Wl,-z,now -fstack-protector -rdynamic -Wl,-export-dynamic -Wl,-E  -Wl,-Bsymbolic-functions -Wl,-z,relro -Wl,-z,now -Wl,--as-needed -o vim   -lgtk-x11-2.0 -lgdk-x11-2.0 -lpangocairo-1.0 -latk-1.0 -lcairo -lgdk_pixbuf-2.0 -lgio-2.0 -lpangoft2-1.0 -lpango-1.0 -lgobject-2.0 -lglib-2.0 -lfontconfig -lfreetype -lSM -lICE -lXpm -lXt -lX11 -lXdmcp -lSM -lICE  -lm -ltinfo -lnsl  -lselinux  -lacl -lattr -lgpm -ldl  -L/usr/lib -llua5.2 -Wl,-E  -fstack-protector-strong -L/usr/local/lib  -L/usr/lib/x86_64-linux-gnu/perl/5.26/CORE -lperl -ldl -lm -lpthread -lcrypt  -L/usr/lib/python3.6/config-3.6m-x86_64-linux-gnu -lpython3.6m -lpthread -ldl -lutil -lm -L/usr/lib/x86_64-linux-gnu -ltcl8.6 -ldl -lz -lpthread -lm -lruby-2.5 -lpthread -lgmp -ldl -lcrypt -lm
```

--- a/python3/vdebug/ui/vimui.py
+++ b/python3/vdebug/ui/vimui.py
@@ -351,6 +351,8 @@ class SourceWindow(interface.Window):
     file = None
     pointer_sign_id = '6145'
     breakpoint_sign_id = '6146'
+    has_sign_priority = vim.vvars['version'] > 801 \
+        or vim.funcs.has('nvim-0-4-0')
 
     def focus(self):
         vim.command("1wincmd w")
@@ -388,8 +390,10 @@ class SourceWindow(interface.Window):
     def place_pointer(self, line):
         log.Log("Placing pointer sign on line "+str(line), log.Logger.INFO)
         self.remove_pointer()
-        vim.command('sign place %s name=current priority=99 line=%s file=%s'
-                    % (self.pointer_sign_id, line, self.file))
+        signstr = 'sign place %s name=current line=%s file=%s'
+        if self.has_sign_priority:
+            signstr = 'sign place %s name=current priority=99 line=%s file=%s'
+        vim.command(signstr % (self.pointer_sign_id, line, self.file))
 
     def remove_pointer(self):
         vim.command('sign unplace %s' % self.pointer_sign_id)
@@ -699,6 +703,8 @@ class StackWindow(Window):
     name = "DebuggerStack"
 
     pointer_sign_id = '6147'
+    has_sign_priority = vim.vvars['version'] > 801 \
+        or vim.funcs.has('nvim-0-4-0')
 
     def on_create(self):
         self.command('inoremap <buffer> <cr> <esc>'
@@ -715,8 +721,11 @@ class StackWindow(Window):
     def place_pointer(self, line):
         log.Log("Stack window: placing pointer sign on line "+str(line), log.Logger.INFO)
         self.remove_pointer()
-        vim.command('sign place %s name=current_stack_position priority=99 line=%s'
-                    % (self.pointer_sign_id, line))
+        signstr = 'sign place %s name=current_stack_position line=%s'
+        if self.has_sign_priority:
+            signstr = \
+                'sign place %s name=current_stack_position priority=99 line=%s'
+        vim.command(signstr % (self.pointer_sign_id, line))
 
     def remove_pointer(self):
         vim.command('sign unplace %s' % self.pointer_sign_id)

